v7.c

```diff
@@ -10261,3 +10273,1012 @@ int main(int argc, char *argv[]) {
+  } else if( y % 4 == 0 && y % 100 != 0 ) {
+    return 366;
+  } else if(y % 100 == 0 && y % 400 != 0) {
+    return 365;
+  } else if(y % 400 == 0) {
+    return 366;
+  } else {
+    return 365;
+  }
+}
+
+static etimeint_t ecma_DayFromYear(etimeint_t y) {
+  return 365 * (y-1970) + floor((y-1969)/4) - floor((y-1901)/100)
+                                              + floor((y-1601)/400);
+}
+
+static etimeint_t ecma_TimeFromYear(etimeint_t y) {
+  return msPerDay * ecma_DayFromYear(y);
+}
+
+static int ecma_YearFromTime_s(etime_t t)
+{
+  int first = floor((t / msPerDay) / 366) + 1970,
+              last = floor((t / msPerDay) / 365) + 1970,
+              middle = 0;
+  
+  if (last < first) {
+    int temp = first;
+    first = last;
+    last = temp;
+  }
+  
+  while (last > first) {
+    middle = (last + first) / 2;
+    if (ecma_TimeFromYear(middle) > t) {
+      last = middle - 1;
+    } else {
+      if (ecma_TimeFromYear(middle) <= t) {
+        if (ecma_TimeFromYear(middle + 1) > t) {
+          first = middle;
+          break;
+        }
+        first = middle + 1;
+      }
+    }
+  }
+  
+  return first;
+}
+
+static int ecma_InLeapYear(etime_t t, int year) {
+  (void)t;
+  return ecma_DaysInYear(year) == 366;
+}
+
+static int ecma_DayWithinYear(etime_t t, int year) {
+  return (int)(ecma_Day(t) - ecma_DayFromYear(year));
+}
+
+static int* ecma_getfirstdays(int isleap) {
+  static int sdays[2][MonthsInYear+1] =
+    {{0, 31, 59, 90, 120, 151, 181, 212, 243, 273, 304, 334, 365},
+     {0, 31, 60, 91, 121, 152, 182, 213, 244, 274, 305, 335, 366}};
+  
+  return sdays[isleap];
+}
+
+static int ecma_MonthFromTime(etime_t t, int year) {
+  int *days;
+  etimeint_t dwy = ecma_DayWithinYear(t, year);
+  
+  unsigned int i; int ret = -1;
+  
+  days = ecma_getfirstdays(ecma_InLeapYear(t, year));
+  
+  for(i = 0; i < MonthsInYear; i++) {
+    if(dwy >= days[i] && dwy < days[i+1]) {
+      ret = i;
+      break;
+    }
+  }
+  
+  return ret;
+}
+
+static int ecma_DateFromTime(etime_t t, int year) {
+  int *days;
+  int mft = ecma_MonthFromTime(t, year);
+  int dwy = ecma_DayWithinYear(t, year);
+  
+  if(mft > 11) {
+    return -1;
+  }
+  
+  days = ecma_getfirstdays(ecma_InLeapYear(t, year));
+  
+  return dwy - days[mft] + 1;
+}
+
+static int ecma_WeekDay(etime_t t) {
+  return abs((ecma_Day(t)+4) % 7);
+}
+
+static int ecma_DaylightSavingTA(etime_t t) {
+  time_t time = t / 1000;
+  struct tm tm;
+  memset(&tm, 0, sizeof(t));
+  localtime_r(&time, &tm);
+  if(tm.tm_isdst > 0) {
+    return msPerHour;
+  } else {
+    return 0;
+  }
+}
+
+static int ecma_LocalTZA() {
+  return (int)-gmtoffms;
+}
+
+static etimeint_t ecma_LocalTime(etime_t t) {
+  return t + ecma_LocalTZA() + ecma_DaylightSavingTA(t);
+}
+
+static etimeint_t ecma_UTC(etime_t t) {
+  return t - ecma_LocalTZA() - ecma_DaylightSavingTA(t-ecma_LocalTZA());
+}
+
+static int ecma_HourFromTime(etime_t t) {
+  return abs((etimeint_t)floor(t / msPerHour) % HoursPerDay);
```

Comment:
Better, but not right enough :-) 
A screendump is better than a thousand words:

```
$ ./v7 -e 'd=new Date("1969-12-31T23:59:59.999Z"); d.toISOString()'
"1969-11-31T01:01:01.001Z"
```